src/engines/arps/ARP/screen.cpp

```diff
@@ -11,55 +15,223 @@ namespace otto::engines::arp {
+      case Playmode::chord: playmode_func_ = play_modes::chord; break;
+      case Playmode::random: playmode_func_ = play_modes::random; break;
+    }
+    update_dots(playmode_func_, octavemode_func_, dummy_state, notes, dots_coords);
+    ui::vg::timeline()
+      .apply(&blue_fade_key)
+      .then<ch::RampTo>(1, fadein_time)
+      .then<ch::Hold>(1, hold_time + fadeout_time)
+      .then<ch::RampTo>(0, fadeout_time);
+    ui::vg::timeline()
+      .apply(&blue_fade_value)
+      .then<ch::RampTo>(1, fadein_time)
+      .then<ch::Hold>(1, hold_time)
+      .then<ch::RampTo>(0, fadeout_time);
   }
 
   void Screen::action(itc::prop_change<&Props::octavemode>, OctaveMode om) noexcept
   {
     octavemode_ = to_string(om);
+    // TODO: Move switch to function in arp.hpp
+    switch (om) {
+      case OctaveMode::standard: octavemode_func_ = octave_modes::standard; break;
+      case OctaveMode::octaveupunison: octavemode_func_ = octave_modes::octaveupunison; break;
+      case OctaveMode::fifthunison: octavemode_func_ = octave_modes::fifthunison; break;
+      case OctaveMode::octaveup: octavemode_func_ = octave_modes::octaveup; break;
+      case OctaveMode::doubleoctaveup: octavemode_func_ = octave_modes::doubleoctaveup; break;
+      case OctaveMode::octavedownup: octavemode_func_ = octave_modes::octavedownup; break;
+    }
+    update_dots(playmode_func_, octavemode_func_, dummy_state, notes, dots_coords);
+    ui::vg::timeline()
+      .apply(&green_fade_key)
+      .then<ch::RampTo>(1, fadein_time)
+      .then<ch::Hold>(1, hold_time + fadeout_time)
+      .then<ch::RampTo>(0, fadeout_time);
+    ui::vg::timeline()
+      .apply(&green_fade_value)
+      .then<ch::RampTo>(1, fadein_time)
+      .then<ch::Hold>(1, hold_time)
+      .then<ch::RampTo>(0, fadeout_time);
   }
 
   void Screen::action(itc::prop_change<&Props::note_length>, float nl) noexcept
   {
     note_length_ = nl;
+    ui::vg::timeline()
+      .apply(&red_fade)
+      .then<ch::RampTo>(1, fadein_time)
+      .then<ch::Hold>(1, hold_time)
+      .then<ch::RampTo>(0, fadeout_time);
   }
 
   void Screen::action(itc::prop_change<&Props::subdivision>, int s) noexcept
   {
     subdivision_ = s;
+    ui::vg::timeline()
+      .apply(&yellow_fade)
+      .then<ch::RampTo>(1, fadein_time)
+      .then<ch::Hold>(1, hold_time)
+      .then<ch::RampTo>(0, fadeout_time);
+  }
+
+  void Screen::update_dots(PlayModeFunc playmode_func_,
+                           OctaveModeFunc octavemode_func_,
+                           ArpeggiatorState state,
+                           NoteArray notes,
+                           util::local_vector<std::pair<nvg::Point, bool>, 48>& dots_coords) noexcept
+  {
+    // Calculate notes
+    //
+    dots.clear();
+    state.reset();
+    do {
+      dots.push_back(octavemode_func_(state, notes, playmode_func_));
+    } while (state.count != 0);
+    if (playmode_func_ != play_modes::chord) dots.pop_back();
```

Comment:
```suggestion
    if (dots.size() > 1) dots.pop_back();
```
I'd say this logic is more correct

Reply:
Yeah, that's correct.